Firestore/CMakeLists.txt

```diff
@@ -15,6 +15,53 @@
 cmake_minimum_required(VERSION 2.8.11)
 project(firestore C CXX)
 
+option(ASAN "Build with Address sanitizer (mutually exculsive with TSAN)" OFF)
+# TODO(varconst): msan
+# Memory sanitizer is more complicated:
+# - it requires all dependencies to be compiled with msan enabled (see
+#   https://github.com/google/sanitizers/wiki/MemorySanitizerLibcxxHowTo);
+# - AppleClang doesn't support it.
+option(TSAN "Build with Thread sanitizer (mutually exculsive with ASAN)" OFF)
+option(UBSAN "Build with Undefined behavior sanitizer" OFF)
+
+if (CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
```

Comment:
cmake/CompilerSetup.cmake defines CLANG and GNU with these values, so you could simplify to just `if (CLANG OR GNU)`. But I don't mind this as is.

(Note that the include of CompilerSetup is below.)

Reply:
Done, thanks.

Reply:
Hmm, I can't move `CompilerSetup` here, because it sets compiler options that don't play nice with Abseil. Conversely, it seems that setting sanitizer options below would affect how the dependencies are compiled, and I thought it's best to compile as much stuff with sanitizers enabled as possible. I might be wrong on this, though.

Reply:
We could also factor out compiler detection from `CompilerSetup` into a separate CMake module. Alternatively, we could make the core logic of setting up the compile a function within `CompilerSetup`. That way we can source the file wherever we please and then specifically invoke our compiler settings in the location that makes the best sense.

An extension of that would be to move this setup into there too since this is largely compiler setup stuff anyway,

Reply:
> We could also factor out compiler detection from CompilerSetup into a separate CMake module

It seems to me that it's the least intrusive way, I can do it in one of the follow-ups.